### PR TITLE
Introduced SeekConsumer

### PR DESCRIPTION
**Pull request by cryogenian**
[ch12801] [ch12802] 

I didn't touch codecs and config here, ldjson is selected in FE by default anyways, and I'm not totally sure that `RawKey + Csv` has no sense. The interesting part is `offsetPull`, hopefully comments are good enough

### CODE REVIEW

File: core/src/main/scala/quasar/datasource/kafka/SeekConsumer.scala [hunk 1]
```diff
@@ -79,54 +90,77 @@ class FullConsumer[F[_]: ConcurrentEffect: ContextShift: Timer, K, V](
...
-      stream: Stream[F, Stream[F, CommittableConsumerRecord[F, K, V]]],
+      stream: Stream[F, Stream[F, CCR]],
       endOffsets: Map[TopicPartition, Long])
-      : Stream[F, CommittableConsumerRecord[F, K, V]] = {
+      : Stream[F, CCR] =
     stream
-      .take(endOffsets.size.toLong) // assumes no automatic assignment; needs a filter otherwise
-      .map(_.takeThrough(isNotOffsetLimit(_, endOffsets)))
+      .take(endOffsets.size.toLong)
+      .map(_.through(offsetPipe(endOffsets)))
```
Comment (wemrysi) [inline core/src/main/scala/quasar/datasource/kafka/SeekConsumer.scala:106]: If we had access to the start position of each partition (i.e. where we seeked to), could we
```scala
_.takeThrough(isNotOffsetLimit(_, endOffsets)).filter(_.offset > seekPosition)
```

Do we know what happens if we attempt to seek to an offset that is less than the available range? Will it error? Postion at the beginning of the partition?

File: core/src/main/scala/quasar/datasource/kafka/SeekConsumer.scala [hunk 2]
```diff
@@ -79,54 +90,77 @@ class FullConsumer[F[_]: ConcurrentEffect: ContextShift: Timer, K, V](
...
+      case Some((ccr, tail)) =>
+        val record = ccr.record
+        val topic = record.topic
+        val partition = record.partition
+        val topicPartition = new TopicPartition(topic, partition)
+        val previous = initialOffsets.get(partition).getOrElse(0L) - 1
+        // From endOffsets javadoc: "the offset of the last available message + 1"
+        val end = endOffsets(topicPartition) - 1
+        // No new events since previous push
+        if (end <= previous) {
```
Comment (wemrysi) [inline core/src/main/scala/quasar/datasource/kafka/SeekConsumer.scala:128]: Is this something we could check earlier, when seeking? Or maybe using `consumer.partitionsMapStream`?

Comment (cryogenian) [reply core/src/main/scala/quasar/datasource/kafka/SeekConsumer.scala:128]: `partitionsMapStream` is unavalable, we're tied to `1.1.x` version of fd4s-kafka unfortunately by custom `KafkaConsumer.java` (or maybe other `*.java` file). With `partitionsMapStream` we definitely could check that inner streams are unnecessary and don't do `+ 1` offset hack.

File: core/src/main/scala/quasar/datasource/kafka/SeekConsumer.scala (hunk 1 above)
Comment (cryogenian) [reply core/src/main/scala/quasar/datasource/kafka/SeekConsumer.scala:106]: No, we don't have `seekPosition` here unfortunately, since we don't have `partition` and can't access `initialOffsets`. If `seek` argument is negative it throws an exception.

Comment (wemrysi) [reply core/src/main/scala/quasar/datasource/kafka/SeekConsumer.scala:106]: Indeed, we don't have access currently, but we could adjust the signature and pass in the `initialOffsets` to use in the `.filter` which I think would allow us to avoid `offsetPull`.

Comment (cryogenian) [reply core/src/main/scala/quasar/datasource/kafka/SeekConsumer.scala]: Thanks! I've just realized that there won't be any halts, since we take one element more, and `.filter` will work :tada:

**Comment by cryogenian** [comment]
+ Switched to `Stream.filter` 

Please, note that in the rare case where `endOffsets` are lesser than previous offsets this might halt.

File: core/src/main/scala/quasar/datasource/kafka/SeekConsumer.scala [hunk 3]
```diff
@@ -34,22 +34,34 @@ import scala.collection.immutable.SortedSet
...
 
-  override def fetch(topic: String): Resource[F, Stream[F, Byte]] = {
+  override def fetch(topic: String): Resource[F, (Offsets, Stream[F, Byte])] = {
     consumerResource[F].using(settings) evalMap { consumer =>
       for {
         endOffsets <- assignNonEmptyPartitionsForTopic(consumer, topic)
+        _ <- endOffsets.toList.traverse_ { case (p, end) =>
+          // `w - 1` because if `w` then in case if there is no new records
+          // CCR stream halts
+          initialOffsets.get(p.partition()).traverse_(w => consumer.seek(p, w - 1))
```
Comment (wemrysi) [inline core/src/main/scala/quasar/datasource/kafka/SeekConsumer.scala:54]: Do we know what happens if we try and seek to a position < beginOffset or > endOffset? Seems like we might want to detect that and error if so? Being outside the partition range would seem to imply we're out of sync and probably missed some data.

We could have the strategy configurable so the user can decide if we should just pickup from the current beginning/end of the partition when out of range (depending on which bound we're outside of) or error.

File: core/src/main/scala/quasar/datasource/kafka/SeekConsumer.scala [hunk 4]
```diff
@@ -34,22 +34,34 @@ import scala.collection.immutable.SortedSet
...
-  import FullConsumer.topicPartitionOrder
+  import SeekConsumer.topicPartitionOrder
 
   val F: ConcurrentEffect[F] = ConcurrentEffect[F]
 
-  override def fetch(topic: String): Resource[F, Stream[F, Byte]] = {
+  override def fetch(topic: String): Resource[F, (Offsets, Stream[F, Byte])] = {
     consumerResource[F].using(settings) evalMap { consumer =>
       for {
         endOffsets <- assignNonEmptyPartitionsForTopic(consumer, topic)
```
Comment (wemrysi) [inline core/src/main/scala/quasar/datasource/kafka/SeekConsumer.scala:50]: If we separate obtaining non-empty partitions from assigning them to the consumer, we could detect which ones actually have new records and only consume from them, allowing us to avoid the off-by-one seek and the `filter`.

**Comment by wemrysi** [comment]
@cryogenian I think we need to ensure we don't block, I think we can with more introspection of offsets, as mentioned in my comments. If there is still a chance after that, we should interrupt the results if no data is seen after some timeout.

File: core/src/main/scala/quasar/datasource/kafka/SeekConsumer.scala (hunk 3 above)
Comment (cryogenian) [reply core/src/main/scala/quasar/datasource/kafka/SeekConsumer.scala:54]: `beginOffset`s are `Map(whatever -> 0)` because we set seek from begining policy, but yeah, if `w < 0` this would be an error, which is impossible, due to saving offsets, and they're always greater than `0` (not even greater or equal, but greater). 

If I understand correctly `> endOffset` would be waiting for this offset to appear.

File: core/src/test/scala/quasar/datasource/kafka/SeekConsumerSpec.scala [hunk 5]
```diff
@@ -78,33 +78,6 @@ class FullConsumerSpec(implicit ec: ExecutionEnv) extends Specification with Ter
     }
   }
 
-  "isOffsetLimit" >> {
```
Comment (wemrysi) [inline core/src/test/scala/quasar/datasource/kafka/SeekConsumerSpec.scala:81]: Can we restore these?

Comment (cryogenian) [reply core/src/test/scala/quasar/datasource/kafka/SeekConsumerSpec.scala:81]: Done

**Review by wemrysi — COMMENTED** [review]
Using the `initialOffsets` simplified things a bit 👍 

Can we add some tests to exercise the resume edge-cases (e.g. no partitions with new data, some partitions with new data, all with new data, etc)?

**Comment by cryogenian** [comment]
Added tests

File: it/src/test/scala/quasar/datasource/kafka/KafkaDatasourceITSpec.scala [hunk 6]
```diff
@@ -157,8 +166,81 @@ class KafkaDatasourceITSpec extends Specification {
...
-  }
 
+    "offsets" >> {
+      def config =
+        ("topics" := List("offsetsPartitioned")) ->:
+        ("decoder" := Decoder.rawValue.asJson) ->:
+        baseConfig
+
+      def configAll =
+        ("topics" := List("offsets-all")) ->:
```
Comment (wemrysi) [inline it/src/test/scala/quasar/datasource/kafka/KafkaDatasourceITSpec.scala:177]: Should this be "offsetsAll" to match the script?

**Review by wemrysi — COMMENTED** [review]
Thanks for adding tests! Do the new topics start off empty? If so would it be useful to seed them with some initial data like the existing ones?

File: it/src/test/scala/quasar/datasource/kafka/KafkaDatasourceITSpec.scala (hunk 6 above)
Comment (cryogenian) [reply it/src/test/scala/quasar/datasource/kafka/KafkaDatasourceITSpec.scala:177]: Yep, sorry, forgot to push ¯\_(ツ)_/

**Comment by cryogenian** [comment]
> If so would it be useful to seed them with some initial data like the existing ones?

Not sure how useful is this, added them for just in case :)

**Review by wemrysi — APPROVED** [review]
